chain/chain/src/flat_storage_creator.rs

```diff
@@ -83,19 +171,170 @@ impl FlatStorageShardCreator {
+                        self.remaining_state_parts = None;
+
+                        let mut store_update = chain_store.store().store_update();
+                        if next_start_part_id < num_parts {
+                            // If there are still remaining state parts, switch status to the new range of state parts.
+                            // We will spawn new rayon tasks on the next status update.
+                            let new_status = FetchingStateStatus {
+                                part_id: next_start_part_id,
+                                num_parts_in_step,
+                                num_parts,
+                            };
+                            debug!(target: "chain", %shard_id, %block_hash, ?new_status);
+                            store_helper::set_fetching_state_status(
+                                &mut store_update,
+                                shard_id,
+                                new_status,
+                            );
+                        } else {
+                            // If all parts were fetched, we can start catchup.
+                            debug!(target: "chain", %shard_id, %block_hash, "Finished fetching state");
+                            store_helper::remove_fetching_state_status(&mut store_update, shard_id);
+                            store_helper::start_catchup(&mut store_update, shard_id);
+                        }
+                        store_update.commit()?;
+
+                        Ok(())
+                    }
+                }
+            }
+            FlatStorageStateStatus::CatchingUp => {
+                let store = self.runtime_adapter.store();
+                let old_flat_head = store_helper::get_flat_head(store, shard_id).unwrap();
+                let mut flat_head = old_flat_head.clone();
+                let chain_final_head = chain_store.final_head()?;
+                let mut merged_delta = FlatStateDelta::default();
+
+                // Merge up to 50 deltas of the next blocks until we reach chain final head.
+                // TODO: consider merging 10 deltas at once to limit memory usage
+                for _ in 0..50 {
+                    let height = chain_store.get_block_height(&flat_head).unwrap();
+                    if height > chain_final_head.height {
+                        panic!("New flat head moved too far: new head = {flat_head}, height = {height}, final block height = {}", chain_final_head.height);
+                    }
+                    if height == chain_final_head.height {
+                        break;
+                    }
+                    flat_head = chain_store.get_next_block_hash(&flat_head).unwrap();
+                    let delta =
+                        store_helper::get_delta(store, shard_id, flat_head).unwrap().unwrap();
+                    merged_delta.merge(delta.as_ref());
+                }
+
+                if old_flat_head != flat_head {
```

Comment:
is it possible that old_flat_head == flat_head == final_head?

Reply:
In reality - no, because final head should move forward at least once during `FetchingState` step.
And even if it is the case, it is fine to wait on this step until final head moves forward.